File: Labtwo/GO/pr2.go
package main

import (
	"fmt"
	"strings"
	"unicode"
)

// Проверка валидности имени пользователя
func isValidUsername(username string) bool {
	// Проверка длины
	if len(username) < 6 || len(username) > 30 {
		return false
	}

	// Не может начинаться/заканчиваться точкой
	if strings.HasPrefix(username, ".") || strings.HasSuffix(username, ".") {
		return false
	}

	// Не может содержать две точки подряд
	if strings.Contains(username, "..") {
		return false
	}

	// Проверка допустимых символов
	for _, c := range username {
		if !(unicode.IsLetter(c) || unicode.IsDigit(c) && c != '.' && c != '*' {
			return false
		}
	}

	return true
}

// Нормализация email
func normalizeEmail(email string) string {
	// Поиск разделителя @
	atPos := strings.Index(email, "@")
	if atPos == -1 {
		return ""  // Некорректный email
	}

	// Разделение на части
	username := email[:atPos]
	domain := email[atPos:]

	var normalized strings.Builder  // Для эффективной конкатенации
	ignoreAfterStar := false       // Флаг игнорирования

	// Обработка имени пользователя
	for _, c := range username {
		if c == '*' {
			ignoreAfterStar = true
			continue
		}

		if !ignoreAfterStar && c != '.' {
			normalized.WriteRune(unicode.ToLower(c))  // В нижний регистр
		}
	}

	// Добавляем домен в нижнем регистре
	normalized.WriteString(strings.ToLower(domain))
	return normalized.String()
}

func main() {
	// Ввод данных
	fmt.Println("Enter emails (separated by commas):")
	var input string
	fmt.Scanln(&input)

	// Хранилище уникальных email (map как множество)
	uniqueEmails := make(map[string]struct{})

	// Разделение ввода по запятым
	for _, email := range strings.Split(input, ",") {
		email = strings.TrimSpace(email)  // Обрезка пробелов
		if email == "" {
			continue
		}

		// Поиск @
		atPos := strings.Index(email, "@")
		if atPos != -1 && isValidUsername(email[:atPos]) {
			normalized := normalizeEmail(email)
			if normalized != "" {
				uniqueEmails[normalized] = struct{}{}  // Добавление в множество
			}
		}
	}

	// Вывод количества уникальных email
	fmt.Printf("Number of unique emails: %d\n", len(uniqueEmails))
}

File: Labtwo/RUBY/pr2.rb
def is_valid_username(username)
  # Проверка длины
  return false if username.length < 6 || username.length > 30
  # Не может начинаться/заканчиваться точкой
  return false if username.start_with?('.') || username.end_with?('.')
  # Не может содержать две точки подряд
  return false if username.include?('..')
  # Проверка допустимых символов через регулярное выражение
  username.match?(/^[a-zA-Z0-9.*]+$/)
end

def normalize_email(email)
  # Поиск @
  at_pos = email.index('@')
  return nil unless at_pos

  # Разделение на части
  username = email[0...at_pos]
  domain = email[at_pos..-1]

  normalized = ''
  ignore_after_star = false

  # Обработка каждого символа
  username.each_char do |c|
    if c == '*'
      ignore_after_star = true
      next
    end

    # Добавляем символ если не точка и не после звездочки
    normalized += c.downcase unless ignore_after_star || c == '.'
  end

  # Возвращаем нормализованный email
  normalized + domain.downcase
end

# Основная программа
print "Enter emails (separated by commas): "
input = gets.chomp

# Множество для хранения уникальных email
unique_emails = Set.new

# Разделение ввода по запятым
input.split(',').each do |email|
  email = email.strip  # Удаление пробелов
  next if email.empty?

  # Поиск @
  at_pos = email.index('@')
  next unless at_pos

  # Проверка валидности и нормализация
  username = email[0...at_pos]
  if is_valid_username(username)
    normalized = normalize_email(email)
    unique_emails.add(normalized) if normalized  # Добавление в множество
  end
end

# Вывод результата
puts "Number of unique emails: #{unique_emails.size}"
